### ISSUE TITLE
[Consensus] State Sync - Define Block By Block Interface

### ISSUE THREAD

**Issue by Olshansk** [pokt-network/pocket#362]
## Objective

Update the interface in `consensus/state_sync/state_sync.go` so it is well-defined enough for the implementation to be picked up by anyone.

## Origin Document

This is a follow-up to #125 per the comments & discussion in #342.

The list of changes this encompasses includes (but is not limited to):
- Remove callback functions and use the application-specific bus for everything
- Adding protobuf definitions
- Splitting the different `Run` functions into separate modules
- Updating terminology if applicable

## Goals

- Finalize the block-by-block state sync design
- Make the interface clear enough such that it is implementable by anyone

## Deliverable

- [ ] Updated interface internal to the consensus module
- [ ] Updated interface in the shared directory
- [ ] Updated protobufs where applicable

## Non-goals / Non-deliverables

- Implement State Sync

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)
- [ ] Update any relevant global documentation & references
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)

## Testing Methodology

- **All tests**: `make test_all`
- **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @Olshansk

**Comment by jessicadaugherty** [comment]
Follow up ticket: Implement Block by Block design (assign to Gokhan)

**Comment by jessicadaugherty** [comment]
Have Gokhan review this ticket to give input/feedback on interface design

**Comment by Olshansk** [comment]
@gokutheengineer Just a heads up that I plan to tackle this ticket in the coming week.

**Comment by jessicadaugherty** [comment]
Thanks @Olshansk please keep us informed about your progress to determine if the implementation will be feasible by 12/14 when iteration 6 begins, and if not, what the next best priority is. 

We have https://github.com/pokt-network/pocket/issues/336 and you already said that you want to take the read.me. 

What about handing off this research https://github.com/pokt-network/pocket/issues/283 to @gokutheengineer?

**Comment by Olshansk** [comment]
> What about handing off this research https://github.com/pokt-network/pocket/issues/283 to @gokutheengineer?

I think understanding the utility context management requires a little more context (no pun intended) before @gokutheengineer should take it on.

> We have https://github.com/pokt-network/pocket/issues/336 and you already said that you want to take the read.me.

@gokutheengineer will help with the consensus readme as he familiarizes himself with the consensus codebase.

There is also a side-task (no issue tracking it yet) that'll involve refactoring how the consensus module and pacemaker interact.

> Thanks @Olshansk please keep us informed about your progress to determine if the implementation will be feasible by 12/14 when iteration 6 begins, and if not, what the next best priority is.

I want finish this in this sprint but do plan to start working on it. At the very least, we can define parts of it (e.g. advertising height and block) which can be implemented independently of the other state sync parts while those interfaces are being defined.

**tl;dr Will focus on defining interfaces to advertise height + block first, which can be done in parallel as a separate issue.**

**Comment by jessicadaugherty** [comment]
> At the very least, we can define parts of it (e.g. advertising height and block) which can be implemented independently of the other state sync parts while those interfaces are being defined.

Thanks @Olshansk! We have this WIP drafts planned for Iteration 7:

![Image](https://user-images.githubusercontent.com/55156619/206926555-c99ca1a6-8a5b-41c7-9172-9fb522f71090.png)

Do you want "request download blocks" moved to Iteration 6 and https://github.com/pokt-network/pocket/issues/336 removed? Or keep as is?

**Comment by Olshansk** [comment]
> > At the very least, we can define parts of it (e.g. advertising height and block) which can be implemented independently of the other state sync parts while those interfaces are being defined.
> 
> Thanks @Olshansk! We have this WIP drafts planned for Iteration 7:
> 
> ![Image](https://user-images.githubusercontent.com/55156619/206926555-c99ca1a6-8a5b-41c7-9172-9fb522f71090.png)
> 
> Do you want "request download blocks" moved to Iteration 6 and #336 removed? Or keep as is?

Sorry for the late response. Design is up for review now (iteration 6): https://github.com/pokt-network/pocket/pull/413

In iteration 7, we'll only work on implementing the advertisement of block & height.
